File: composeApp/src/commonMain/kotlin/app/rigel/bridge/Bridges.kt
package app.rigel.bridge

import co.touchlab.kermit.Logger
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlin.coroutines.resume

/**
 * Suspend wrappers over the callback-style bridge interfaces.
 * Throws [IllegalStateException] when a bridge is not registered (Swift app startup missing).
 */
object Bridges {
    private const val TAG = "Bridges"

    private fun <T> requireBridge(name: String, value: T?): T =
        requireNotNull(value) { "$name bridge is not registered — call RigelBridgeFactory.register() at Swift app startup" }

    suspend fun ssdpSearch(searchTargets: List<String>, timeoutMs: Int): List<SsdpDevice> {
        val bridge = requireBridge("Discovery", RigelBridgeFactory.discovery)
        return suspendCancellableCoroutine { cont ->
            bridge.ssdpSearch(searchTargets, timeoutMs) { devices ->
                if (cont.isActive) cont.resume(devices)
            }
        }
    }

    suspend fun probe(url: String, headers: Map<String, String>): Pair<ProbeResult?, String?> {
        val bridge = requireBridge("Probe", RigelBridgeFactory.probe)
        return suspendCancellableCoroutine { cont ->
            bridge.probe(url, headers) { result, error ->
                if (cont.isActive) cont.resume(result to error)
            }
        }
    }

    suspend fun startHlsSession(
        sessionId: String,
        sourceUrl: String,
        headers: Map<String, String>,
        mode: String,
        onError: (String) -> Unit,
    ): Pair<String?, String?> {
        val bridge = requireBridge("Transcode", RigelBridgeFactory.transcode)
        return suspendCancellableCoroutine { cont ->
            bridge.startHlsSession(
                sessionId,
                sourceUrl,
                headers,
                mode,
                onReady = { path, error ->
                    if (cont.isActive) cont.resume(path to error)
                },
                onError = onError,
            )
        }
    }

    fun stopHlsSession(sessionId: String) {
        RigelBridgeFactory.transcode?.stopHlsSession(sessionId)
            ?: Logger.w(TAG) { "Transcode bridge not registered; stopHlsSession ignored" }
    }

    suspend fun startHttpServer(): Pair<Long, String?> {
        val bridge = requireBridge("HttpServer", RigelBridgeFactory.httpServer)
        return suspendCancellableCoroutine { cont ->
            bridge.start { port, error ->
                if (cont.isActive) cont.resume(port to error)
            }
        }
    }

    fun stopHttpServer() {
        RigelBridgeFactory.httpServer?.stop()
    }

    fun lanBaseUrl(): String? = RigelBridgeFactory.httpServer?.lanBaseUrl()
}

File: iosApp/iosApp/Bridge/Bridges.swift
import Foundation
import ComposeApp

// MARK: - Kotlin interface conformances (Nuvio MPVPlayerBridge pattern)

final class RigelDiscoveryBridge: NSObject, DiscoveryBridge {
    func ssdpSearch(searchTargets: [String], timeoutMs: Int32, onResult: @escaping ([SsdpDevice]) -> Void) {
        Ssdp.search(searchTargets: searchTargets, timeoutMs: Int(timeoutMs), onResult: onResult)
    }
}

final class RigelProbeBridge: NSObject, ProbeBridge {
    func probe(url: String, headers: [String: String], onResult: @escaping (ProbeResult?, String?) -> Void) {
        DispatchQueue.global(qos: .userInitiated).async {
            let (result, error) = RigelProbe.probe(url: url, headers: headers)
            DispatchQueue.main.async { onResult(result, error) }
        }
    }
}

final class RigelTranscodeBridge: NSObject, TranscodeBridge {
    func startHlsSession(
        sessionId: String,
        sourceUrl: String,
        headers: [String: String],
        mode: String,
        startOffsetMs: Int64,
        subtitleTracks: [SubtitleTrack],
        onReady: @escaping (String?, String?) -> Void,
        onError: @escaping (String) -> Void
    ) {
        RigelHlsExporter.startSession(
            sessionId: sessionId,
            sourceUrl: sourceUrl,
            headers: headers,
            mode: mode,
            startOffsetMs: startOffsetMs,
            subtitleTracks: subtitleTracks,
            onReady: onReady,
            onError: onError
        )
    }

    func stopHlsSession(sessionId: String) {
        RigelHlsExporter.stopSession(sessionId: sessionId)
    }
}

final class RigelHttpServerBridge: NSObject, HttpServerBridge {
    private let server = RigelHttpServer(documentRoot: RigelHttpServer.proxyRootURL())

    func start(onStarted: @escaping (KotlinLong, String?) -> Void) {
        server.start { port, error in
            onStarted(KotlinLong(longLong: Int64(port ?? -1)), error)
        }
    }

    func stop() {
        server.stop()
    }

    func lanBaseUrl() -> String? {
        guard let ip = RigelHttpServer.lanIPv4(), let port = server.port else { return nil }
        return "http://\(ip):\(port)"
    }
}

// MARK: - Registration (called from Swift app startup)

enum BridgeRegistry {
    static func register() {
        RigelBridgeFactory.shared.register(
            discovery: RigelDiscoveryBridge(),
            probe: RigelProbeBridge(),
            transcode: RigelTranscodeBridge(),
            httpServer: RigelHttpServerBridge()
        )
        PlayerBridgeFactory.shared.register(bridge: RigelPlayerBridge())
        RendererBridgeFactory.shared.register(bridge: RigelRendererBridge())
        ChromecastBridgeFactory.shared.register(bridge: RigelChromecastBridge())
    }
}
